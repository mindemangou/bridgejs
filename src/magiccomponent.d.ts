export type Connected=({element,props}:{element:HTMLElement,props:unknown})=>void;

export type Disconnected=( ({element}:{element:HTMLElement})=>void )|null;

export type GlobaleElementConstructor=(connected:Connected,disconnected:Disconnected)=>CustomElementConstructor;


declare const registerCustomElement: (tagName: string, ClassRef: CustomElementConstructor) => void;

export declare const define: (
  name: string,
  connected: Connected,
  disconnected?: Disconnected | null
) => Promise<void>;

export declare const getPath: (
  query: Record<string, string>,
  fragment: string
) => string;

export declare const reload: (
  name: string,
  query?: Record<string, string>,
  fragment?: string
) => void;

export declare const getProps: (element: HTMLElement) => Record<string, any>;

export declare const redirect: (
  url: string,
  headers?: object
) => Promise<boolean | void>;

export declare const config: (options: {
  redirect: boolean;
  loader?: { enable: boolean; color?: string };
}) => Promise<void>;