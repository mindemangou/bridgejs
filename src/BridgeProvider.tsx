import { createContext, PropsWithChildren, useContext } from "react"

const BridgeContext=createContext<unknown>({})

type BrigeProviderType=PropsWithChildren<{
    data:unknown
}>
export const BridgeProvider=({children,data}:BrigeProviderType)=> {

    return <BridgeContext value={data}>
                {children}
            </BridgeContext>
}

export const useBridgeData=<T extends {}>()=> {

    const data=useContext(BridgeContext) as unknown as T

    return data
}