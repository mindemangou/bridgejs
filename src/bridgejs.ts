import getCustomElementConstructor, { type Connected, type Disconnected } from "./globaleElement.ts"


const  registerCustomElement=(tagName:string, ClassRef:CustomElementConstructor)=>{
    if (!customElements.get(tagName)) {
      customElements.define(tagName, ClassRef);
    }
  }
 


export const define=(name:string,connected:Connected,disconnected:Disconnected=null)=> {
    
    registerCustomElement(name,getCustomElementConstructor(connected,disconnected,name)) 
}


export const reloadElement=(names:string[])=> {

  for (const name of names) {

    const element=document.querySelector(name)

    if(element) {
        element?.setAttribute('reload',name)
    }

  }
  
}


  