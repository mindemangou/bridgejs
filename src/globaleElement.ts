import camelCase from "lodash.camelcase"


export type Connected=({element,props}:{element:HTMLElement,props:unknown})=>void;

export type Disconnected=( ({element}:{element:HTMLElement})=>void )|null;

export type GlobaleElementConstructor=(connected:Connected,disconnected:Disconnected,name:string)=>CustomElementConstructor


 const getGlobalElementConstructor:GlobaleElementConstructor=(connected:Connected,disconnected:Disconnected,name:string)=> {

        
    class GlobalElement extends HTMLElement{

        static observedAttributes = ["reload"];

        static originalElement:Node|null=null;

        constructor() {
            
            super();
          }

        connectedCallback() {
            const props=this.getProps()

            connected({element:this,props})
        }

        disconnectedCallBack() {

            if(disconnected) {
                disconnected({element:this})
            }
            
        }

       async attributeChangedCallback(_name:string, _oldValue:string, newValue:string) {
            const {ajax}=await import('htmx.org')
            ajax('GET',location.toString(),{target:newValue,select:newValue,swap:'outerHTML'})
        }

        private getProps() {

            const map=new Map()

            for (const element of this.attributes) {
                map.set(camelCase(element.name),element.value)
            }

            const template=this.querySelector('template')

            if(template) {
              //Get template content
              const content=template?.content.textContent?.trim()

              //Parse json
              const parseContent=content?JSON.parse(content):' '

              map.set('data',parseContent)
            }
            
            const props=Object.fromEntries(map)

            return props
        }

    
    }
    
    return GlobalElement

 }


 export default getGlobalElementConstructor
