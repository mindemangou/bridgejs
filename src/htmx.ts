import {config} from "./magiccomponents";

config({redirect:true,loader:{enable:true,color:'#a3d'}})

//on('htmx:xhr:progress',()=>console.log('progress'))

/* window.addEventListener("popstate",()=> {
    //history.go()
    const items=localStorage.getItem('htmx-history-cache')

    if(items) {
        const parsItem=JSON.parse(items) as any[]
        const path=document.location.pathname
        const element=parsItem.find((el)=> {
            return el.url===path
        })
        document.body.innerHTML+="<a href='/about.php?id=4' hx-boost='true'>test</a>"
        process(document.body)
        console.log(element.content)
    }
  }); */