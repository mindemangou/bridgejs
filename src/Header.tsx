import { useBridgeData } from "./BridgeProvider"

export const Header=()=> {
    const {age}=useBridgeData()
    return <>
        <header>
            <span>title is ?? {age}</span>
        </header>
    </>
}