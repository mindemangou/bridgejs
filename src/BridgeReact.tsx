import { createContext, type PropsWithChildren, useCallback, useContext, useState } from "react"
import {ajax,trigger} from 'htmx.org'
import { getProps } from "./magiccomponents"

const BridgeContext=createContext({})



type BrigeProviderType=PropsWithChildren<{
    init:any
}>

export const BridgeProvider=({children,init}:BrigeProviderType)=> {
    return <BridgeContext value={{...init}}>
                {children}
            </BridgeContext>
}


export const useBridgeData=()=> {
    
    const contextData=useContext(BridgeContext) as any
    
    console.log(contextData)
    
    const [state,setState]=useState({contextData,processing:false})

    const refresh=useCallback(()=> {
            
        setState((oldState)=>({...oldState,processing:true}))

        const template=document.createElement('template')
    
        const tagName=contextData.tagName 
    
        template.id=tagName
        
        document.body.appendChild(template)
    
        ajax('GET',location.toString(),{target:`#${tagName}`,select:`${tagName}`,swap:'innerHTML'}).then(()=> {
            
            const element=template.firstElementChild as HTMLElement
    
            const newdata=getProps(element)
    
            setState(()=>({contextData:newdata,processing:false}))
    
        }).then(()=>template.remove())
    
    },[])

    const sendMessage=useCallback((name:string,data:{[key in string]:any})=> {
            
        const element=document.querySelector(name)
        
        if(element) {
            
            trigger(element,'incoming_message',data)
        }

    },[])


    return {...state.contextData,processing:state.processing,refresh,sendMessage}
}

export const useMessage=()=> {
    
    const data=useContext(BridgeContext) as any
    
    const [state,setState]=useState({})
    
    const element=document.querySelector(data.tagName)

    element?.addEventListener('incoming_message',(evt:{detail:object})=> {
       
        setState(()=>{
            return evt?.detail
        })
        
    })

    return state
}
