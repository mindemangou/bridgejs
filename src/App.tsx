import { createRoot } from "react-dom/client"
import { define } from "./bridgejs"
import { Header } from "./Header"
import { PropsWithChildren, useEffect } from "react"
import { BridgeProvider, useBridgeData } from "./BridgeProvider"

/* type PropsType={
    age:string
} */

const App=()=> {

    const {age,data}=useBridgeData<{age:string,data:{version:string}}>()

    useEffect(()=> {
        console.log(data)
    });

    return <>
        <h1>HOME my age is {age??'??'}</h1>
        <Header />
    </>
}

define('app-app',({element,props})=> {

    createRoot(element).render(
        <BridgeProvider data={props}>
            <App/>
        </BridgeProvider>
        )
})
