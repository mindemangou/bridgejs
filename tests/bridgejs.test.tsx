import {test,expect, vi} from 'vitest'
import {render} from '@testing-library/react'
import {BridgeProvider} from '../src/BridgeProvider'
import {Hello} from './Hello'
import { Browser, BrowserErrorCaptureEnum } from 'happy-dom'

import {
  getByLabelText,
  getByText,
  getByTestId,
  queryByTestId,
  waitFor,
  screen,
  
} from '@testing-library/dom'
import { define } from '../src/bridgejs'
import React from 'react'

function getExampleDOM() {

  const div = document.createElement('main')
  div.innerHTML = `
    <hello-hello userid='10' ></hello-hello>
  `
  return div
}


test('essai',async ()=> {

  const browser = new Browser({ settings: { errorCapture: BrowserErrorCaptureEnum.processLevel } });
const page = browser.newPage();
const body=page.mainFrame.document.body
body.innerHTML= `
<hello-hello userid='10' ></hello-hello>
`


// Aborts all ongoing operations and closes the browser
await browser.close();

  const dom=body
 
  define('hello-hello',({element,props})=> {

    render(<BridgeProvider data={props}><Hello/></BridgeProvider>,{
      container: document?.body?.appendChild(dom)
    })

    
  })

  console.log(body.innerHTML)

  
 /*  const hello=document.body.querySelector('hello-hello')
  console.log(hello)
  expect(hello?.innerHTML).toBe('<h1>Hi! John doe</h1>') */
  
})


/* test('examples of some things', async () => {
  const famousProgrammerInHistory = 'Ada Lovelace'
  const container = getExampleDOM()

  // Get form elements by their label text.
  // An error will be thrown if one cannot be found (accessibility FTW!)
  const input = getByLabelText(container, 'Username')
  input.value = famousProgrammerInHistory

  // Get elements by their text, just like a real user does.
  getByText(container, 'Print Username').click()

  await waitFor(() =>
    expect(queryByTestId(container, 'printed-username')).toBeTruthy(),
  )

  // getByTestId and queryByTestId are an escape hatch to get elements
  // by a test id (could also attempt to get this element by its text)
  expect(getByTestId(container, 'printed-username').innerHTML).toBe(
    famousProgrammerInHistory,
  )
  // jest snapshots work great with regular DOM nodes!
  expect(container).toMatchSnapshot()
}) */