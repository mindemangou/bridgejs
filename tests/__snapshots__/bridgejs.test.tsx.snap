// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`essai 1`] = `
<main>
  
    
  <hello-hello
    userid="10"
  />
  
  
</main>
`;
